pkg/testutil/store.go

```diff
@@ -0,0 +1,125 @@
+package testutil
+
+import (
+	"context"
+	"errors"
+	"io"
+	"testing"
+
+	"github.com/improbable-eng/thanos/pkg/store/storepb"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/tsdb/chunkenc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// StoreClient is test gRPC store API client.
+type StoreClient struct {
+	Values map[string][]string
+
+	RespSet []*storepb.SeriesResponse
+}
+
+func (s *StoreClient) Info(ctx context.Context, req *storepb.InfoRequest, _ ...grpc.CallOption) (*storepb.InfoResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "not implemented")
+}
+
+func (s *StoreClient) Series(ctx context.Context, req *storepb.SeriesRequest, _ ...grpc.CallOption) (storepb.Store_SeriesClient, error) {
+	return &StoreSeriesClient{ctx: ctx, respSet: s.RespSet}, nil
+}
+
+func (s *StoreClient) LabelNames(ctx context.Context, req *storepb.LabelNamesRequest, _ ...grpc.CallOption) (*storepb.LabelNamesResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "not implemented")
+}
+
+func (s *StoreClient) LabelValues(ctx context.Context, req *storepb.LabelValuesRequest, _ ...grpc.CallOption) (*storepb.LabelValuesResponse, error) {
+	return &storepb.LabelValuesResponse{Values: s.Values[req.Label]}, nil
+}
+
+// StoreSeriesClient is test gRPC storeAPI series client.
+type StoreSeriesClient struct {
+	// This field just exist to pseudo-implement the unused methods of the interface.
+	storepb.Store_SeriesClient
+	ctx     context.Context
+	i       int
+	respSet []*storepb.SeriesResponse
+}
+
+func (c *StoreSeriesClient) Recv() (*storepb.SeriesResponse, error) {
+	if c.i >= len(c.respSet) {
+		return nil, io.EOF
+	}
+	s := c.respSet[c.i]
+	c.i++
+
+	return s, nil
+}
+
+func (c *StoreSeriesClient) Context() context.Context {
+	return c.ctx
+}
+
+// StoreSeriesServer is test gRPC storeAPI series server.
+type StoreSeriesServer struct {
+	// This field just exist to pseudo-implement the unused methods of the interface.
+	storepb.Store_SeriesServer
+	ctx context.Context
+
+	SeriesSet []storepb.Series
+	Warnings  []string
+}
+
+func NewStoreSeriesServer(ctx context.Context) *StoreSeriesServer {
+	return &StoreSeriesServer{ctx: ctx}
+}
+
+func (s *StoreSeriesServer) Send(r *storepb.SeriesResponse) error {
+	if r.GetWarning() != "" {
+		s.Warnings = append(s.Warnings, r.GetWarning())
+		return nil
+	}
+
+	if r.GetSeries() == nil {
+		return errors.New("no seriesSet")
+	}
+	s.SeriesSet = append(s.SeriesSet, *r.GetSeries())
+	return nil
+}
+
+func (s *StoreSeriesServer) Context() context.Context {
+	return s.ctx
+}
```

Comment:
If you really want to have this extracted somehow, maybe a different proposal. One issue with this is that it only returns the specific item. If you want to extend errors or make the calls have any side-effects (like counting or recording requests) this won't work and will either be extended or another implementation pops up.

What works nicely I think is a `Func*` type, where you can set functions as attributes and it simply calls them. You can set the functions inline in your tests so it's always clear what happens under the hood. But you don't need to out a new type everywhere.

Example:

```go
package storepb

type StoreSeriesFuncServer struct {
    SeriesFunc func(*storepb.SeriesResponse) error
    ContextFunc func() context.Context
}

func (s *StoreSeriesFuncServer) Send(r *storepb.SeriesResponse) error {
    return s.SeriesFunc(r)
}
```

In tests:

```go
s := StoreSeriesFuncServer{
    SendFunc: func(r *storepb.SeriesResponse) error {
        // record, count, or do whatever the test actually needs.
    }
}

// Use s as a StoreSeriesServer.
```

Reply:
For anything below this comment, I see no value in having this here rather than in the single package that uses it.

Reply:
+1